demo/repository.go

```diff
@@ -0,0 +1,530 @@
+	}
+
+	item.lockedByTransactionID = r.transactionID
+
+	return item.event, nil
+}
+
+func (r *EventRepository) FindOldestUnprocessed(ctx context.Context) (*events.Event, error) {
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+
+	slog.DebugContext(ctx, "finding oldest unprocessed event")
+
+	var oldestUnprocessedEvent *events.Event
+	for _, item := range r.items {
+		if item.event.ProcessedAt != nil {
+			continue
+		}
+
+		if oldestUnprocessedEvent == nil || item.event.Timestamp.Before(oldestUnprocessedEvent.Timestamp) {
+			oldestUnprocessedEvent = item.event
+		}
+	}
+
+	if oldestUnprocessedEvent == nil {
+		return nil, events.ErrNotFound
+	}
+
+	return oldestUnprocessedEvent, nil
+}
+
+func (r *EventRepository) FindUnprocessed(ctx context.Context, limit int) ([]*events.Event, error) {
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+
+	slog.DebugContext(ctx, "finding unprocessed events", "limit", limit)
+
+	var unprocessedEvents []*events.Event
+	for _, item := range r.items {
+		if len(unprocessedEvents) >= limit {
+			break
+		}
+
+		if item.event.ProcessedAt != nil {
+			continue
+		}
+
+		unprocessedEvents = append(unprocessedEvents, item.event)
+	}
+
+	return unprocessedEvents, nil
+}
+
+func (r *EventRepository) Transaction(ctx context.Context, fn func(txRepo events.EventRepository) error) error {
+	var txRepo *EventRepository
+	if r.transactionID == 0 {
+		txRepo = &EventRepository{
+			items:                r.items,
+			mutex:                r.mutex,
+			transactionID:        r.transactionIDCounter.Add(1),
+			transactionIDCounter: r.transactionIDCounter,
+		}
+
+		defer func() {
+			r.mutex.Lock()
+			defer r.mutex.Unlock()
+
+			for _, item := range r.items {
+				item.lockedByTransactionID = 0
+			}
+		}()
+	} else {
+		txRepo = r
+	}
+
+	slog.DebugContext(ctx, "starting event repository transaction", "transactionId", txRepo.transactionID)
+
+	err := fn(txRepo)
+
+	slog.DebugContext(ctx, "finished event repository transaction", "transactionId", txRepo.transactionID, "error", err)
+
+	return err
+}
+
+func (r *EventRepository) Update(ctx context.Context, event *events.Event) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	slog.DebugContext(ctx, "updating event", "eventId", event.ID, "transactionId", r.transactionID)
+
+	item, found := r.items[event.ID]
+	if !found {
+		return errors.New("there is no event with the same ID in the collection")
+	}
+
+	if item.lockedByTransactionID != 0 && item.lockedByTransactionID != r.transactionID {
+		return errors.New("this event is locked by another transaction")
+	}
+
+	item.lockedByTransactionID = 0
+
+	item.event = event
+
+	return nil
+}
+
+type HandlerRequestRepositoryItem struct {
+	handlerRequest        *events.HandlerRequest
+	lockedByTransactionID int64
+}
+
+type HandlerRequestRepository struct {
+	items                map[string]*HandlerRequestRepositoryItem
+	mutex                *sync.RWMutex
+	transactionID        int64
+	transactionIDCounter *atomic.Int64
```

Comment:
Realizing this is demo code, but these should probably be stored on the struct rather than a pointer:

```suggestion
	mutex                sync.RWMutex
	transactionID        int64
	transactionIDCounter atomic.Int64
```

Reply:
It was necessary to have these as pointers so that the same mutex could be shared by the different copies of the repository (the `Transaction` method would clone the repository), but this is no longer necessary now that the mutexes have been moved into a struct shared by all the store+repository copies.